REPLACE INTO panics on FULLTEXT-indexed table when old and new content share a word at the same position

Hi — we ran into this while upgrading to pick up the #10729 fix (which is working great, thank you).

`REPLACE INTO` on a table with a FULLTEXT index panics when the old and new text share a word at the same position. `ON DUPLICATE KEY UPDATE` hits the same issue. Completely different content works fine.

**Minimal repro** (embedded driver, inline SQL):

```go
package main

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"path/filepath"

	_ "github.com/dolthub/driver"
)

func main() {
	dir, _ := os.MkdirTemp("", "ft-repro-*")
	defer os.RemoveAll(dir)
	dsn := fmt.Sprintf("file://%s?commitname=t&commitemail=t@t&database=d",
		filepath.ToSlash(dir))
	db, _ := sql.Open("dolt", dsn)
	defer db.Close()
	ctx := context.Background()
	conn, _ := db.Conn(ctx)
	defer conn.Close()

	conn.ExecContext(ctx, "CREATE DATABASE IF NOT EXISTS d")
	conn.ExecContext(ctx, "USE d")
	conn.ExecContext(ctx, "CREATE TABLE t (id INT PRIMARY KEY, body LONGTEXT NOT NULL)")
	conn.ExecContext(ctx, "CREATE FULLTEXT INDEX ft ON t(body)")

	conn.ExecContext(ctx, "INSERT INTO t VALUES (1, 'hello world')")

	// Panics — "hello" appears at position 0 in both old and new content:
	conn.ExecContext(ctx, "REPLACE INTO t VALUES (1, 'hello earth')")

	// This would work — no shared words:
	// conn.ExecContext(ctx, "REPLACE INTO t VALUES (1, 'goodbye earth')")
}
```

```
panic: impossible conversion: string cannot be converted to int

goroutine 1 [running]:
github.com/dolthub/dolt/go/store/prolly/tree.convInt(...)
    store/prolly/tree/prolly_fields.go:699
```

**What we think is happening:**

`MultiTableEditor.Insert` runs the fulltext secondary insert before the primary table insert. When the old and new text share a word at the same position, the fulltext position table's INSERT raises a `PrimaryKeyViolation` — the key `[word, parent_pk, position]` already exists from the original row. The REPLACE loop in `insertIter.Next` (insert.go:192) catches this as a generic `UniqueKeyError`, but `ue.Existing` is the **position table's** row (e.g. `["hello", 1, 0]`), not the main table's row (`[1, "hello world"]`). It passes this wrong-shaped row to `Delete`, and the main table's prolly index writer panics trying to interpret a word string as an INT primary key.

`ON DUPLICATE KEY UPDATE` hits the same root cause — the position table PK violation gets caught at insert.go:209 and `ue.Existing` is passed to `handleOnDuplicateKeyUpdate`, which tries to update the main table with the position table's row shape.

The trigger condition is: old and new body text share at least one word at the same position. For example:
- `'hello world'` → `'hello earth'` — panics ("hello" at position 0 in both)
- `'hello world'` → `'goodbye earth'` — works (no shared words)
- `'func foo'` → `'func bar'` — panics ("func" at position 0 in both)

**Possible fix direction:**

The position table Insert in `fulltext_editor.go:334` propagates PK violations:
```go
if err = index.Position.Editor.Insert(ctx, positionRow); err != nil {
    return err
}
```

But the doccount table Insert right below it (line 358) already suppresses them:
```go
if err = index.DocCount.Editor.Insert(ctx, docCountRow); err != nil &&
    !sql.ErrPrimaryKeyViolation.Is(err) && !sql.ErrUniqueKeyViolation.Is(err) && !sql.ErrDuplicateEntry.Is(err) {
    return err
}
```

If the position table Insert tolerated duplicates the same way, the fulltext secondary insert would succeed, and the primary table insert would then raise the correct PK violation from the main table — which the REPLACE loop knows how to handle.

**What works as a workaround:**

`UPDATE` and `DELETE` + `INSERT` both work fine with the same content.

**Versions:**
- `dolthub/dolt/go v0.40.5-0.20260415052817-250fcf57cd3d`
- `dolthub/go-mysql-server v0.20.1-0.20260412215059-d7fc9477f4b7`
- `dolthub/driver v1.84.2-0.20260415010027-81fe8bae47a0`
- Go 1.26, Linux amd64

We haven't tested with `dolt sql-server` — only the embedded driver. Happy to test further if helpful.

--- a/enginetest/queries/fulltext_queries.go
+++ b/enginetest/queries/fulltext_queries.go
@@ -1205,4 +1205,31 @@ var FulltextTests = []ScriptTest{
 			},
 		},
 	},
+	{
+		// https://github.com/dolthub/dolt/issues/10882
+		Name: "REPLACE INTO/INSERT...ON DUPLICATE KEY UPDATE when words are in the same position",
+		SetUpScript: []string{
+			"CREATE TABLE t (id INT PRIMARY KEY, body LONGTEXT NOT NULL)",
+			"CREATE FULLTEXT INDEX ft ON t(body)",
+			"INSERT INTO t VALUES (1, 'hello world')",
+		},
+		Assertions: []ScriptTestAssertion{
+			{
+				Query:    "REPLACE INTO t VALUES (1, 'hello earth')",
+				Expected: []sql.Row{{types.OkResult{RowsAffected: 2}}},
+			},
+			{
+				Query:    "select * from t",
+				Expected: []sql.Row{{1, "hello earth"}},
+			},
+			{
+				Query:    "INSERT INTO t VALUES (1, 'hello world') ON DUPLICATE KEY UPDATE body = VALUES(body)",
+				Expected: []sql.Row{{types.OkResult{RowsAffected: 2}}},
+			},
+			{
+				Query:    "select * from t",
+				Expected: []sql.Row{{1, "hello world"}},
+			},
+		},
+	},
 }

--- a/sql/fulltext/fulltext_editor.go
+++ b/sql/fulltext/fulltext_editor.go
@@ -331,6 +331,10 @@ func (editor TableEditor) Insert(ctx *sql.Context, row sql.Row) error {
 			positionRow[0] = word
 			positionRow = append(positionRow, keyCols...)
 			positionRow = append(positionRow, position)
+			// TODO: In some places, for example `REPLACE INTO` and `INSERT...ON DUPLICATE UPDATE`, we use the information
+			//  inside UniqueKeyError to perform the replace/update. However, the information in the UniqueKeyError
+			//  generated from an Insert on index.Position.Editor does not match the primary table's schema.
+			//  https://github.com/dolthub/dolt/issues/10882#issuecomment-4255176383
 			if err = index.Position.Editor.Insert(ctx, positionRow); err != nil {
 				return err
 			}

--- a/sql/fulltext/fulltext_editor.go
+++ b/sql/fulltext/fulltext_editor.go
@@ -331,11 +331,8 @@ func (editor TableEditor) Insert(ctx *sql.Context, row sql.Row) error {
 			positionRow[0] = word
 			positionRow = append(positionRow, keyCols...)
 			positionRow = append(positionRow, position)
-			// TODO: In some places, for example `REPLACE INTO` and `INSERT...ON DUPLICATE UPDATE`, we use the information
-			//  inside UniqueKeyError to perform the replace/update. However, the information in the UniqueKeyError
-			//  generated from an Insert on index.Position.Editor does not match the primary table's schema.
-			//  https://github.com/dolthub/dolt/issues/10882#issuecomment-4255176383
-			if err = index.Position.Editor.Insert(ctx, positionRow); err != nil {
+			// TODO: write to this table only once, rather than ignoring duplicate errors
+			if err = index.Position.Editor.Insert(ctx, positionRow); err != nil && !sql.ErrPrimaryKeyViolation.Is(err) && !sql.ErrUniqueKeyViolation.Is(err) && !sql.ErrDuplicateEntry.Is(err) {
 				return err
 			}
 		}
@@ -358,7 +355,7 @@ func (editor TableEditor) Insert(ctx *sql.Context, row sql.Row) error {
 			docCountRow[0] = word
 			docCountRow = append(docCountRow, keyCols...)
 			docCountRow = append(docCountRow, wordDocCount)
-			//TODO: write to this table only once, rather than ignoring duplicate errors
+			// TODO: write to this table only once, rather than ignoring duplicate errors
 			if err = index.DocCount.Editor.Insert(ctx, docCountRow); err != nil && !sql.ErrPrimaryKeyViolation.Is(err) && !sql.ErrUniqueKeyViolation.Is(err) && !sql.ErrDuplicateEntry.Is(err) {
 				return err
 			}

--- a/sql/fulltext/multi_editor.go
+++ b/sql/fulltext/multi_editor.go
@@ -93,12 +93,21 @@ func (editor MultiTableEditor) StatementComplete(ctx *sql.Context) error {
 
 // Insert implements the interface sql.TableEditor.
 func (editor MultiTableEditor) Insert(ctx *sql.Context, row sql.Row) error {
+	// Any error generated from inserting on a primary table should take precedence over errors generated from
+	// inserting on a secondary table.
+	if err := editor.primary.Insert(ctx, row); err != nil {
+		return err
+	}
 	for _, secondary := range editor.secondaries {
+		// TODO: In some places, for example `REPLACE INTO` and `INSERT...ON DUPLICATE UPDATE`, we use the information
+		//  inside UniqueKeyError to perform the replace/update. However, the information in the UniqueKeyError
+		//  generated from an Insert on a secondary table does not match the primary table's schema.
+		//  https://github.com/dolthub/dolt/issues/10882#issuecomment-4255176383
 		if err := secondary.Insert(ctx, row); err != nil {
 			return err
 		}
 	}
-	return editor.primary.Insert(ctx, row)
+	return nil
 }
 
 // Update implements the interface sql.TableEditor.

--- a/sql/fulltext/multi_editor.go
+++ b/sql/fulltext/multi_editor.go
@@ -93,21 +93,16 @@ func (editor MultiTableEditor) StatementComplete(ctx *sql.Context) error {
 
 // Insert implements the interface sql.TableEditor.
 func (editor MultiTableEditor) Insert(ctx *sql.Context, row sql.Row) error {
-	// Any error generated from inserting on a primary table should take precedence over errors generated from
-	// inserting on a secondary table.
-	if err := editor.primary.Insert(ctx, row); err != nil {
-		return err
-	}
 	for _, secondary := range editor.secondaries {
 		// TODO: In some places, for example `REPLACE INTO` and `INSERT...ON DUPLICATE UPDATE`, we use the information
 		//  inside UniqueKeyError to perform the replace/update. However, the information in the UniqueKeyError
-		//  generated from an Insert on a secondary table does not match the primary table's schema.
+		//  generated from an Insert on a secondary table does always not match the primary table's schema.
 		//  https://github.com/dolthub/dolt/issues/10882#issuecomment-4255176383
 		if err := secondary.Insert(ctx, row); err != nil {
 			return err
 		}
 	}
-	return nil
+	return editor.primary.Insert(ctx, row)
 }
 
 // Update implements the interface sql.TableEditor.

--- a/sql/rowexec/insert.go
+++ b/sql/rowexec/insert.go
@@ -189,6 +189,8 @@ func (i *insertIter) Next(ctx *sql.Context) (returnRow sql.Row, returnErr error)
 					return nil, sql.NewWrappedInsertError(row, err)
 				}
 
+				// TODO: For multitables, UniqueKeyError.Existing might not be the correct row if the error is coming
+				//  from a secondary table. https://github.com/dolthub/dolt/issues/10882#issuecomment-4255176383
 				ue := err.(*errors.Error).Cause().(sql.UniqueKeyError)
 				if err = i.replacer.Delete(ctx, ue.Existing); err != nil {
 					i.rowSource.Close(ctx)
@@ -206,6 +208,8 @@ func (i *insertIter) Next(ctx *sql.Context) (returnRow sql.Row, returnErr error)
 		if err := i.inserter.Insert(ctx, row); err != nil {
 			if (sql.ErrPrimaryKeyViolation.Is(err) || sql.ErrUniqueKeyViolation.Is(err)) &&
 				i.onDupKeyUpdateExprs.HasUpdates() {
+				// TODO: For multitables, UniqueKeyError.Existing might not be the correct row if the error is coming
+				//  from a secondary table. https://github.com/dolthub/dolt/issues/10882#issuecomment-4255176383
 				if uniqueKeyError, ok := err.(*errors.Error).Cause().(sql.UniqueKeyError); ok {
 					return i.handleOnDuplicateKeyUpdate(ctx, uniqueKeyError.Existing, row)
 				}

